[compile] Enable sequence parallelism matching w/o custom ops enabled

angelayi opened the pull request:
## Purpose

Based on https://github.com/vllm-project/vllm/pull/24604, modified sequence-parallelism pass to do custom op matching w/o needing to enable the custom op

## Test Plan

`pytest -sv tests/compile/test_sequence_parallelism.py`

## Performance numbers

I did some benchmarking with the command on H100 w/o flashinfer

```
VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_STANDALONE_COMPILE=1 VLLM_LOGGING_LEVEL=DEBUG vllm bench latency --model=nvidia/Llama-3.3-70B-Instruct-FP8 --output-len 1 --input-len 8192 --batch-size 1 --tensor-parallel-size 8 --load-format dummy --num_iters_warmup 5 --num_iters 15 -O '{"level": 3, "use_inductor_graph_partition": false, "splitting_ops":[], "cudagraph_mode": "FULL", }' --no-enable-prefix-caching
```

while varying
* `"pass_config": {"enable_async_tp": true, "enable_sequence_parallelism": true}` vs. `"pass_config": {"enable_async_tp": false, "enable_sequence_parallelism": false}`
* `"custom_ops":["+quant_fp8", "+rms_norm"]` vs. `"custom_ops":[]`

<img width="718" height="1046" alt="image" src="https://github.com/user-attachments/assets/349f5da6-2965-48b5-9fca-7abca413ed59" />

angelayi committed:
[compile] Enable sequence parallelism matching w/o custom ops enabled

Signed-off-by: angelayi <yiangela7@gmail.com>

angelayi committed:
[compile] Fix rmsnorm

Signed-off-by: angelayi <yiangela7@gmail.com>
Co-authored-by: Luka Govedič <lgovedic@redhat.com>

ProExpertProg commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -43,172 +43,157 @@
+    "test_model_cls, custom_ops",
+    [
+        # (TestAllReduceRMSNormModel, "+rms_norm"),
+        # (TestAllReduceRMSNormModel, "-rms_norm"),
```
Comment?

ProExpertProg commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -310,38 +300,28 @@ def sequence_parallelism_pass_on_test_model(
+        backend.check_after_ops(model.ops_in_model_after())
 
-        # check if the functionalization pass is applied
+        print(backend.graph_post_pass)
```
Remove print?

ProExpertProg commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -10,109 +12,39 @@
-    """Base helper for RMSNorm and RMSNorm + Quantization functionalization."""
+def get_first_out_wrapper(fn):
+    @functools.wraps(fn)
+    def wrapper(*args):
```
Does this work? I thought that during tracing the pattern matching tracer will think that `args` is a single parameter

angelayi replied:
yes! updated the test to assert the number of all_reduce/all_gather ops in the graph!

ProExpertProg reviewed (commented):
Thanks for taking this on! Could you just add me as a co-author on one of the commits?

ProExpertProg reviewed (approved):
@cascade812 could you take a look at this please?

cascade812 commented:
> @cascade812 could you take a look at this please?

Sure!

angelayi committed:
Add e2e tests

Co-authored-by: Luka Govedič <lgovedic@redhat.com>

Signed-off-by: angelayi <yiangela7@gmail.com>

cascade812 commented:
@angelayi  I have below error if not specify `custom_ops=["+rms_norm"]`
```
torch._inductor.exc.InductorError: RuntimeError: The size of tensor a (s72) must match the size of tensor b ((s72//2)) at non-singleton dimension 0)
```

ProExpertProg commented on tests/compile/test_fusions_e2e.py:
```diff
@@ -125,6 +145,11 @@ def test_attn_quant(
+        pytest.skip(
+            "Known bug for fp4 fusion & inductor partition: "
+            "https://github.com/vllm-project/vllm/issues/26988"
+        )
```
#26988 was fixed
```suggestion
```

ProExpertProg reviewed (approved):
We no longer have to skip the FP4 tests!

angelayi committed:
Fix test_sequence_parallel + FP8

Signed-off-by: angelayi <yiangela7@gmail.com>

ProExpertProg committed:
more robust fp8 check

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>

ProExpertProg committed:
Merge branch 'main' into sp_custom_op

angelayi committed:
disable async tp tests

Signed-off-by: angelayi <yiangela7@gmail.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into sp_custom_op

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
fix SP test with +rms_norm for piecewise graph

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
Better handling of custom rms norm, fix unit test, reduce warnings

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
Fix e2e fusion numbers

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>

ProExpertProg committed:
Merge branch 'refs/heads/main' into sp_custom_op

ProExpertProg committed:
Reorganize tests to fix failures and improve coverage:
- Remove FI backend from llama4
- Move kv scales test to non-optional
- Fix custom op selection
- properly skip fp4 tests

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
Merge branch 'main' into sp_custom_op

ProExpertProg commented on .buildkite/test-pipeline.yaml:
```diff
@@ -475,10 +475,11 @@ steps:
     # Limit to no custom ops to reduce running time
     # Wrap with quotes to escape yaml and avoid starting -k string with a -
-  - "pytest -v -s compile/test_fusions_e2e.py -k 'TRITON and -quant_fp8'"
+  - "pytest -v -s compile/test_fusions_e2e.py -k 'TRITON and not +quant_fp8'"
```
```suggestion
  - "pytest -v -s compile/test_fusions_e2e.py -k 'TRITON and not +quant_fp8' and not Llama-4"
```

ProExpertProg committed:
Merge branch 'main' into sp_custom_op

ProExpertProg committed:
Fix CI

- remove llama4 from sm89
- fix llama4 asynctp count

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>

ProExpertProg committed:
Merge branch 'main' into sp_custom_op

ProExpertProg committed:
Fix pre-commit

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into sp_custom_op

ProExpertProg committed:
FI broken on Blackwell for llama4, broken on Hopper with fp8 kvcache

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ProExpertProg committed:
Typo in CI

Signed-off-by: ProExpertProg <lgovedic@redhat.com>

ZJY0516 commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -310,38 +303,41 @@ def sequence_parallelism_pass_on_test_model(
+        ]
+        for op in model.ops_in_model_before():
+            num_op = len([pre_op for pre_op in pre_ops if pre_op == op])
+            assert num_op == 4
```
```suggestion
        for op in model.ops_in_model_before():
            assert backend.op_count(op, before=True) == 4
```

ZJY0516 commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -310,38 +303,41 @@ def sequence_parallelism_pass_on_test_model(
+        ]
+        for op in model.ops_in_model_after():
+            num_op = len([post_op for post_op in post_ops if post_op == op])
+            assert num_op == 4
```
```suggestion
        for op in model.ops_in_model_after():
            assert backend.op_count(op, before=False) == 4
```

ZJY0516 reviewed (commented):
I think this would make the logic clearer and more straightforward

angelayi committed:
Skip llama-4 in distributed tests

Signed-off-by: angelayi <yiangela7@gmail.com>

ProExpertProg committed:
fix SP count for qwen

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>

ProExpertProg commented on tests/compile/test_fusions_e2e.py:
```diff
@@ -68,15 +97,24 @@ class ModelBackendTestCase(NamedTuple):
+                allreduce_fusion=97,
+                # TODO(luka) perhaps avoid SP for MoE allreduce? 49
+                sequence_parallel=97,
+                async_tp=48,  # mlp is moe, no fusion there
```
```suggestion
                async_tp=96,  # MLP is MoE, half the fusions of dense
```

ProExpertProg commented on tests/compile/test_fusions_e2e.py:
```diff
@@ -68,15 +97,24 @@ class ModelBackendTestCase(NamedTuple):
+            matches=Matches(
+                attention_fusion=0,
+                allreduce_fusion=97,
+                # TODO(luka) perhaps avoid SP for MoE allreduce? 49
```
```suggestion
```

ProExpertProg committed:
change comment

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>

ProExpertProg committed:
remove unnecessary todo

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>
